In [1]:
import pandas as pd
import numpy as np
from sklearn.metrics import precision_recall_curve, auc, log_loss

# Data understanding

## Collect initial data 

**Task: Collect initial data**\
Acquire within the project the data (or access to the data) listed in the
project resources. This initial collection includes data loading if necessary
for data understanding. For example, if you apply a specific tool for data
understanding, it makes perfect sense to load your data into this tool.
This effort possibly leads to initial data preparation steps.
Note: if you acquire multiple data sources, integration is an additional
issue, either here or in the later data preparation phase

**Output: Initial data collection report**\
List the dataset (or datasets) acquired, together with their locations
within the project, the methods used to acquire them and any problems
encountered. Record problems encountered and any solutions achieved
to aid with future replication of this project or with the execution of
similar future projects

In [2]:
file = 'training_sample.tsv'

In [3]:
column_names = ["text_tokens", "hashtags", "tweet_id", "present_media", "present_links", "present_domains",\
                "tweet_type", "language", "tweet_timestamp", "engaged_with_user_id", "engaged_with_user_follower_count",\
               "engaged_with_user_following_count", "engaged_with_user_is_verified", "engaged_with_user_account_creation",\
               "engaging_user_id", "engaging_user_follower_count", "engaging_user_following_count", "engaging_user_is_verified",\
               "engaging_user_account_creation", "engaged_follows_engaging", "reply_timestamp", "retweet_timestamp", "retweet_with_comment_timestamp", "like_timestamp"]

In [4]:
df = pd.read_csv(file, header=None, names=column_names, delimiter='\x01')

In [5]:
pd.set_option('display.max_columns', None)
print(df.shape)
display(df.head())

(80425, 24)


text_tokens  \
0  101\t56898\t137\t174\t63247\t10526\t131\t3197\...   
1  101\t102463\t10230\t10105\t21040\t10169\t12811...   
2  101\t56898\t137\t11255\t22037\t10263\t168\t111...   
3  101\t13073\t28757\t106\t100\t14120\t131\t120\t...   
4  101\t3460\t1923\t6632\t2824\t30368\t2179\t1881...   

                           hashtags                          tweet_id  \
0                               NaN  3C21DCFB8E3FEC1CB3D2BFB413A78220   
1                               NaN  3D87CC3655C276F1771752081423B405   
2  DB32BD91C2F1B37BE700F374A07FBC61  3701848B96AA740528A2B0E247777D7D   
3                               NaN  18176C6AD2871729384062F073CCE94D   
4                               NaN  AF11AF01F842E7F120667B7B0B38676D   

  present_media                     present_links  \
0         Video                               NaN   
1           NaN  BB422AA00380E45F312FD2CAA75F4960   
2           NaN  2423BA02A75DB2189335DDC3FB6B74A1   
3         Video                               NaN   
4           NaN                               NaN   

                    present_domains tweet_type  \
0                               NaN    Retweet   
1  92D397F8E0F1E77B36B8C612C2C51E23   TopLevel   
2  6D323BE93766E79BE423FAC5C28BE39B    Retweet   
3                               NaN   TopLevel   
4                               NaN      Quote   

                           language  tweet_timestamp  \
0  76B8A9C3013AE6414A3E6012413CDC3B       1581467323   
1  D3164C7FBCF2565DDF915B1B3AEFB1DC       1580975391   
2  22C448FF81263D4BAF2A176145EE9EAD       1581257232   
3  D3164C7FBCF2565DDF915B1B3AEFB1DC       1581164292   
4  22C448FF81263D4BAF2A176145EE9EAD       1581233650   

               engaged_with_user_id  engaged_with_user_follower_count  \
0  D1AA2C85FA644D64346EDD88470525F2                               737   
1  4DC65AC7BD963DE1F7617C047C33DE99                          52366425   
2  5C671539CB41B9807E209349B101E9FF                               988   
3  70B900BE17416923D1E236A38798F202                           1228134   
4  E94C0E9E8494F3D603F9D1A5C5242E3D                                73   

   engaged_with_user_following_count  engaged_with_user_is_verified  \
0                                706                          False   
1                               2383                           True   
2                                167                          False   
3                               5413                          False   
4                                299                          False   

   engaged_with_user_account_creation                  engaging_user_id  \
0                          1403069820  000046C8606F1C3F5A7296222C88084B   
1                          1230139136  00006047187D0D18598EF12A650E1DAC   
2                          1530094483  0000648BAA193AE4C625DDF789B57172   
3                          1378699943  000071667F50BAFEA722A8E8284581E5   
4                          1549054499  00007745A6EE969F1A0F44B10DC17671   

   engaging_user_follower_count  engaging_user_following_count  \
0                           131                           2105   
1                            22                             50   
2                           251                            719   
3                            18                             58   
4                           268                            526   

   engaging_user_is_verified  engaging_user_account_creation  \
0                      False                      1573978269   
1                      False                      1340673962   
2                      False                      1456473671   
3                      False                      1378427564   
4                      False                      1252294800   

   engaged_follows_engaging  reply_timestamp  retweet_timestamp  \
0                     False              NaN                NaN   
1                     False              NaN                NaN   
2       

## Describe data 

**Task: Describe data**\
Examine the “gross” or “surface” properties of the acquired data and
report on the results.

**Output: Data description report**\
Describe the data which has been acquired, including: the format of
the data, the quantity of data, for example number of records and fields
in each table, the identities of the fields and any other surface features
of the data which have been discovered. Does the data acquired satisfy
the relevant requirements?

| Feature category    | Feature name                 | Feature dtype | Feature description                                                                           |
|---------------------|------------------------------|---------------|-----------------------------------------------------------------------------------------------|
| User features       | userId                       | string        | User identifier                                                                               |
| User features       | follower count               | int           | Number of followers of the user                                                               |
| User features       | following count              | int           | Number of accounts this user is following                                                     |
| User features       | is verified                  | bool          | Is the account verified?                                                                      |
| User features       | account creation             | timestamp     | in ms int Unix timestamp (in seconds) of the creation time of the account                     |
| Tweet features      | tweetId                      | string        | Tweet identifier                                                                              |
| Tweet features      | presentMedia                 | list[string]  | Tab-separated list of media types;  media type can be in (Photo, Video, Gif)                  |
| Tweet features      | presentLinks                 | list[string]  | Tab-separated list of links included in the tweet                                             |
| Tweet features      | presentDomains               | list[string]  | Tab-separated list of domains (e.g. twitter.com) included in the tweet                        |
| Tweet features      | tweetType                    | string        | Tweet type, can be either Retweet, Quote, Reply, or Toplevel                                  |
| Tweet features      | language                     | string        | Identifier corresponding to inferred language of the tweet                                    |
| Tweet features      | tweet timestamp              | int           | Unix timestamp, in seconds of the creation time of the Tweet                                  |
| Tweet features      | tweet tokens                 | list[int]     | Ordered list of Bert ids corresponding to Bert tokenization of Tweet text                     |
| Tweet features      | tweet hashtags               | list[string]  | Tab-separated list of hashtags present in the tweet                                           |
| Engagement features | reply engagement timestamp   | int           | Unix timestamp, in seconds, of the Reply engagement if one exists.                            |
| Engagement features | retweet engagement timestamp | int           | Unix timestamp, in seconds, of the Retweet engagement if one exists.                          |
| Engagement features | quote engagement timestamp   | int           | Unix timestamp, in seconds, of the Quote engagement if one exists.                            |
| Engagement features | like engagement timestamp    | int           | Unix timestamp, in seconds, of the Like engagement if one exists.                             |
| Engagement features | engageeFollowsEngager        | bool          | Does the account of the engaged tweet author follow the account that has made the engagement? |

## Explore data 


**Task: Explore data**\
This task tackles the data mining questions, which can be addressed
using querying, visualization and reporting. These include: distribution
of key attributes, for example the target attribute of a prediction task;
relations between pairs or small numbers of attributes; results of
simple aggregations; properties of significant sub-populations; simple
statistical analyses. These analyses may address directly the data mining goals; they may also contribute to or refine the data description
and quality reports and feed into the transformation and other data
preparation needed for further analysis.


**Output: Data exploration report\
Describe results of this task including first findings or initial hypothesis and their impact on the remainder of the project. If appropriate,
include graphs and plots, which indicate data characteristics or lead
to interesting data subsets for further examination.

## Verify data quality

**Task: Verify data quality**\
Examine the quality of the data, addressing questions such as: is the
data complete (does it cover all the cases required)? Is it correct or
does it contain errors and if there are errors how common are they?
Are there missing values in the data? If so how are they represented,
where do they occur and how common are they?


**Output: Data quality report**\
List the results of the data quality verification; if quality problems
exist, list possible solutions. Solutions to data quality problems
generally depend heavily on both data and business knowledge.

In [6]:
display(df)

text_tokens  \
0      101\t56898\t137\t174\t63247\t10526\t131\t3197\...   
1      101\t102463\t10230\t10105\t21040\t10169\t12811...   
2      101\t56898\t137\t11255\t22037\t10263\t168\t111...   
3      101\t13073\t28757\t106\t100\t14120\t131\t120\t...   
4      101\t3460\t1923\t6632\t2824\t30368\t2179\t1881...   
...                                                  ...   
80420  101\t56898\t137\t14796\t13711\t17617\t10161\t1...   
80421  101\t56898\t137\t11885\t11273\t40154\t10206\t1...   
80422  101\t22800\t10531\t10124\t28780\t104939\t10230...   
80423  101\t18249\t112\t187\t169\t16745\t26133\t117\t...   
80424  101\t78879\t10671\t10266\t19493\t16213\t10411\...   

                               hashtags                          tweet_id  \
0                                   NaN  3C21DCFB8E3FEC1CB3D2BFB413A78220   
1                                   NaN  3D87CC3655C276F1771752081423B405   
2      DB32BD91C2F1B37BE700F374A07FBC61  3701848B96AA740528A2B0E247777D7D   
3                                   NaN  18176C6AD2871729384062F073CCE94D   
4                                   NaN  AF11AF01F842E7F120667B7B0B38676D   
...                                 ...                               ...   
80420  FC7321735734C2FC8A3CAE30D266CD71  533F80610C8C2F4345517986B5BB58E5   
80421  0FE0A5F06FA20E3C2CDE7F65ACA0046C  DAC3216BB2DC4747BB2CCBA6D253A308   
80422                               NaN  5F9EFD38F96180EAB6BAA74481C0E6FE   
80423                               NaN  8CB10325EAAD5E121E686EF222B8598C   
80424                               NaN  E19A3CD27AFBB44DF363C1A364FCE35B   

      present_media                     present_links  \
0             Video                               NaN   
1               NaN  BB422AA00380E45F312FD2CAA75F4960   
2               NaN  2423BA02A75DB2189335DDC3FB6B74A1   
3             Video                               NaN   
4               NaN                               NaN   
...             ...                               ...   
80420           NaN                               NaN   
80421           NaN                               NaN   
80422           NaN                               NaN   
80423         Photo                               NaN   
80424           NaN                               NaN   

                        present_domains tweet_type  \
0                                   NaN    Retweet   
1      92D397F8E0F1E77B36B8C612C2C51E23   TopLevel   
2      6D323BE93766E79BE423FAC5C28BE39B    Retweet   
3                                   NaN   TopLevel   
4                                   NaN      Quote   
...                                 ...        ...   
80420                               NaN    Retweet   
80421                               NaN    Retweet   
80422                               NaN      Quote   
80423                               NaN   TopLevel   
80424                               NaN   TopLevel   

                               language  tweet_timestamp  \
0      76B8A9C3013AE6414A3E6012413CDC3B       1581467323   
1      D3164C7FBCF2565DDF915B1B3AEFB1DC       1580975391   
2      22C448FF81263D4BAF2A176145EE9EAD       1581257232   
3      D3164C7FBCF2565DDF915B1B3AEFB1DC       1581164292   
4      22C448FF81263D4BAF2A176145EE9EAD       1581233650   
...                                 ...              ...   
80420  D3164C7FBCF2565DDF915B1B3AEFB1DC       1581212842   
80421  06D61DCBBE938971E1EA0C38BD9B5446       1581533061   
80422  D3164C7FBCF2565DDF915B1B3AEFB1DC       1581119848   
80423  D3164C7FBCF2565DDF915B1B3AEFB1DC       1581052325   
80424  ECED8A16BE2A5E8871FD55F4842F16B1       1581350690   

                   engaged_with_user_id  engaged_with_user_follower_count  \
0      D1AA2C85FA644D64346EDD88470525F2                               737   
1      4DC65AC7BD963DE1F7617C047C33DE99                          52366425   
2      5C671539CB41B9807E209349B101E9FF                               988   
3      70B900BE17416923D1E236A38798F

Q: Why do some users have the exactly same timestamp for two types of engagements (e.g. row_id 80424)?\
A: 

# Data Preparation

## Select data

**Task: Select data**\
Decide on the data to be used for analysis. Criteria include relevance
to the data mining goals, quality and technical constraints such as
limits on data volume or data types. Note that data selection covers
selection of attributes (columns) as well as selection of records (rows)
in a table.

**Output: Rationale for inclusion/exclusion**\
List the data to be included/excluded and the reasons for these decisions.

## Clean data

**Task: Clean data**\
Raise the data quality to the level required by the selected analysis
techniques. This may involve selection of clean subsets of the data, the
insertion of suitable defaults or more ambitious techniques such as the
estimation of missing data by modeling.


**Output: Data cleaning report**\
Describe what decisions and actions were taken to address the data
quality problems reported during the verify data quality task of the
data understanding phase. Transformations of the data for cleaning
purposes and the possible impact on the analysis results should be
considered.

## Construct data

**Task: Construct data**\
This task includes constructive data preparation operations such as the
production of derived attributes, entire new records or transformed
values for existing attributes.


**Outputs: Derived attributes**\
Derived attributes are new attributes that are constructed from one or more
existing attributes in the same record. Examples: area = length * width.
Generated records
Describe the creation of completely new records. Example: create
records for customers who made no purchase during the past year.
There was no reason to have such records in the raw data, but for
modeling purposes it might make sense to explicitly represent the fact
that certain customers made zero purchases.

### Enhance dataset with derived columns

## Independent variables
The following new variables were defined in the dataframe:

| Name | Type | Description |
| --- | --- | --- |
| has_media | bool | Indicates that is there any media type related to the tweet or not? (should be omitted later, it was just a helper for further extraction) |
| no_Photo | Integer | number of photo media in the tweet |
| no_Video | Integer | number of video media in the tweet|
| no_GIF | Integer | number of GIF media in the tweet |
| no_Other_media | Integer | (maybe it's impossible) if there is an unknown media type, it will be counted |
| no_Hashtags | Integer | Count the number of hashtags in each tweet |
| no_Domains | Integer | Count the number of domains in each tweet |
| no_Links | Integer | Count the number of links in each tweet |
| no_Words_In_Tweet | Integer | Count the number of words in the tweet |

## Labels
| Name | Type | Description |
| --- | --- | --- |
| is_like | bool | indicates that was there a like by the engaging user on the tweet (decided on that is there timestamp exists or not) |
| is_reply | bool | indicates that was there a reply by the engaging user on the tweet (decided on that is there timestamp exists or not) |
| is_retweet | bool | indicates that was there a retweet by the engaging user on the tweet (decided on that is there timestamp exists or not) |
| is_retweet_with_comment | bool | indicates that was there a retweet with comment by the engaging user on the tweet (decided on that is there timestamp exists or not) |

Is there any media in the cell?

In [7]:
df['has_media'] = df['present_media'].notna()

In [8]:
df[['has_media','present_media']]

has_media present_media
0           True         Video
1          False           NaN
2          False           NaN
3           True         Video
4          False           NaN
...          ...           ...
80420      False           NaN
80421      False           NaN
80422      False           NaN
80423       True         Photo
80424      False           NaN

[80425 rows x 2 columns]

## Count the different media types for every tweet

In [9]:
photoCnt = []
videoCnt = []
gifCnt = []
other = []
for d,b in zip(df['present_media'],  df['has_media']):
    if b is True:
        pht = d.count("Photo")
        video = d.count("Video")
        gif  = d.count("GIF")
        photoCnt.append(pht)
        videoCnt.append(video)
        gifCnt.append(gif)        
        if len(d) != int(pht + video + gif):
             other.append(len(d) - int(pht + video + gif))
        else:
             other.append(0)
    else:
        photoCnt.append(0)
        videoCnt.append(0)
        gifCnt.append(0)
        other.append(0)                       

In [10]:
df['no_Photo'] = photoCnt
df['no_Video'] = videoCnt
df['no_GIF'] = gifCnt
df['no_Other_Media'] = other

In [11]:
df[['present_media','no_Photo', 'no_Video', 'no_GIF', 'no_Other_Media']].head(5)

present_media  no_Photo  no_Video  no_GIF  no_Other_Media
0         Video         0         1       0               4
1           NaN         0         0       0               0
2           NaN         0         0       0               0
3         Video         0         1       0               4
4           NaN         0         0       0               0

## Count the number of hashtags in each tweet

In [12]:
df['no_Hashtags'] = df.apply(lambda x : len(str(x['hashtags']).split()) if str(x['hashtags']) != "nan" else 0,axis=1)

In [13]:
df[['hashtags','no_Hashtags']].head(5)

hashtags  no_Hashtags
0                               NaN            0
1                               NaN            0
2  DB32BD91C2F1B37BE700F374A07FBC61            1
3                               NaN            0
4                               NaN            0

In [14]:
df[df['no_Hashtags'] != 0][['hashtags', 'no_Hashtags']].head(20)

hashtags  no_Hashtags
2                    DB32BD91C2F1B37BE700F374A07FBC61            1
10                   9887C2F9C8FFECE3524054D91E871F84            1
12                   94838D747A18E1270C051B6DDFCAAE0D            1
16                   0EE6C006799F0AFD5655BDC4418B9A07            1
17  CA7AF9A82452F923905E4F8A485DED43\t9E53645466A2...            6
19                   DC3CA8E5056079E402ED35A7B1FDEC23            1
20                   E2774F85FC328CCF3515AF7B12EF18F0            1
21                   9A65AD335F99CB560602444D4F9111C8            1
22  1BBE947EB038868E02C649D57688A100\t7C2090819424...            2
31  CD33508ADA28A80D3A0DD6728CD7A537\t6F5D51E47F9A...            2
34                   B114D402D84DB8F5CCB91F25795F444E            1
36                   B6024D6F02AFAF9BC8ED941F72C228F1            1
37  F56A64A9F426D87BEA76E9A296EE006C\t47FB0777794C...           10
38  FE4B1491A2B93DE49B401E0D1279784E\tFB5A802DF7D5...            3
39  4BDBA4AFD880EC2978072C923091EFB5\t174207A25D4E...            3
40  283E10F8BE4496B4BD7CEB62BD9A277C\tACF41558C81F...            2
41  1D49A2A8F6340E66BDD04A2E0516B260\t459A2A009D78...            3
47  466397B084F7F0E479420BA91EC3F291\tAD8BA8480018...            4
48                   16554C336F78240B4BA2B369D7EE1BAA            1
49                   EF48F0520EB04007A1864B892CB38AD7            1

## Count the number of links in each tweet

In [15]:
df['no_Links'] = df.apply(lambda x : len(str(x['present_links']).split()) if str(x['present_links']) != "nan" else 0,axis=1)

In [16]:
df[df['no_Links'] != 0][['present_links', 'no_Links']].head(20)

present_links  no_Links
1                     BB422AA00380E45F312FD2CAA75F4960         1
2                     2423BA02A75DB2189335DDC3FB6B74A1         1
5                     BB79CD318A68247B64F0E0BE7AFD5A92         1
8                     EF5D0A312E7A9BAEFBEA72A062E5F6CE         1
12                    F28E60037FB10D95E6CF3ADD83D67EA4         1
19                    AABF8A38DBBEE6D8B6FA3650262AC2D4         1
31                    F30118EB13346C80D754EAC2067092D5         1
36                    60DE00805B565526B32FF78F6EF0B9F7         1
37                    6EA7E07DDEDDBB1AAF7C7D9BFC232CAA         1
43                    9B1A416C5A7E6FED4A461F9F2108EB83         1
49   459F45BFE472BA192BC0615493A50BDB\t7ECBAA4765BF...         2
64                    1CB2E9D79689E547B32144E76BFA834F         1
70                    FFD41FE58DCA44A252A10E12DE9209CD         1
75                    0EDA872A7CD8EF9EF41AA3AF5C4DF38E         1
79                    DB7B100FFDA2EE43C3890CF6B7FD3908         1
84                    64437F367209412FA2B7DBA65A99716E         1
89                    6A287C68AA99FE049FDD579A70A86693         1
95                    94A6D5DAFD9DCAC96716D92A9FD4EF5B         1
102                   F8BBF1C5FCE5F850806B20E5191E87C7         1
114                   6B3A4D4CAA6B87A30A55AB9914C2B901         1

In [17]:
df[['present_links', 'no_Links']].head(5)

present_links  no_Links
0                               NaN         0
1  BB422AA00380E45F312FD2CAA75F4960         1
2  2423BA02A75DB2189335DDC3FB6B74A1         1
3                               NaN         0
4                               NaN         0

## Count the number of domains in each tweet

In [18]:
df['no_Domains'] = df.apply(lambda x : len(str(x['present_domains']).split()) if str(x['present_domains']) != "nan" else 0,axis=1)

In [19]:
df[df['no_Domains'] != 0][['present_domains', 'no_Domains']].head(20)

present_domains  no_Domains
1                     92D397F8E0F1E77B36B8C612C2C51E23           1
2                     6D323BE93766E79BE423FAC5C28BE39B           1
5                     3896E26D12C903F0A00B6B1BE9A9BEA3           1
8                     3183ACF54B4022B25B4157B81C174DD5           1
12                    0F510BF067278AAD1ECAECC3380B0162           1
19                    3896E26D12C903F0A00B6B1BE9A9BEA3           1
31                    178B23A1119D4345DA847C10943997F9           1
36                    8F8CD97E2BFE3675BD1333C33D8F662B           1
37                    AA1AD16575EB76EB0CD1D2D951DC1F88           1
43                    7A52A6DD8470FC7770FDF9610909B635           1
49   3896E26D12C903F0A00B6B1BE9A9BEA3\t3896E26D12C9...           2
64                    F4A765D203E9604A20DF6C5362737718           1
70                    097C323141E21EA4941416B3BF4172E0           1
75                    1677479761CC167B998EF821EDC1331B           1
79                    B4A8AFACE5E163157B6128E79958C7E4           1
84                    8F261585DA0A3A006B3AFCC3368BE43A           1
89                    EC3B35626467D4D973FAA96D9101424E           1
95                    6825B061BF551A1617BE63A12C2A3CB6           1
102                   E85D1F68A0B4739BE39EAE4EEBB579D0           1
114                   1F5DE1B702AA81C85EB1D7B3D3029B50           1

In [20]:
df[['present_domains', 'no_Domains']].head(5)

present_domains  no_Domains
0                               NaN           0
1  92D397F8E0F1E77B36B8C612C2C51E23           1
2  6D323BE93766E79BE423FAC5C28BE39B           1
3                               NaN           0
4                               NaN           0

## Count the number of words in the tweet

It is also worth to mention, that there are separators in the text with ID [101], [131], [102] (but at least 2 separator in each tweet) 

In [21]:
df['no_Words_In_Tweet'] = df.apply(lambda x : len(str(x['text_tokens']).split()) if str(x['text_tokens']) != "nan" else 0,axis=1)

In [22]:
df[["text_tokens",'no_Words_In_Tweet']].head(20)

text_tokens  no_Words_In_Tweet
0   101\t56898\t137\t174\t63247\t10526\t131\t3197\...                 31
1   101\t102463\t10230\t10105\t21040\t10169\t12811...                 34
2   101\t56898\t137\t11255\t22037\t10263\t168\t111...                 89
3   101\t13073\t28757\t106\t100\t14120\t131\t120\t...                 22
4   101\t3460\t1923\t6632\t2824\t30368\t2179\t1881...                 31
5   101\t46242\t40751\t161\t100062\t10107\t10114\t...                 45
6   101\t56898\t137\t10192\t11373\t10500\t131\t219...                 13
7   101\t65724\t15619\t22859\t14120\t131\t120\t120...                 20
8   101\t100\t45031\t43804\t10121\t10146\t89387\t1...                 59
9   101\t56898\t137\t26037\t91678\t168\t15734\t737...                 47
10  101\t10725\t14015\t14107\t13028\t10114\t12688\...                 44
11  101\t1925\t11662\t111801\t14244\t2452\t3709\t4...                105
12  101\t10117\t14329\t44871\t10107\t10301\t10893\...                 63
13  101\t35888\t10146\t74176\t55818\t12292\t10121\...                 34
14  101\t11005\t10116\t10106\t84732\t11272\t11904\...                 24
15  101\t11065\t21852\t117\t10549\t10944\t100\t188...                 62
16  101\t56898\t137\t72337\t10858\t25096\t10115\t1...                 46
17  101\t4854\t4097\t1933\t6667\t1921\t89052\t2252...                 87
18  101\t4867\t1980\t8462\t100\t14120\t131\t120\t1...                 22
19  101\t60635\t25474\t10317\t10338\t10891\t11327\...                 43

Average number of words

In [23]:
np.mean(df['no_Words_In_Tweet'])

48.516990985390116

Minimum number of words

In [24]:
np.min(df['no_Words_In_Tweet'])

2

Max number of words

In [25]:
np.max(df['no_Words_In_Tweet'])

255

## Encode what kind of interactions was the engaging user did with the tweet
Decided on timestamp: if there is timestamp, the interaction has happened (1), else no (0)
- Is_reply 
- Is_retweet
- is _retweet_with_comment
- Is_like

In [26]:
df['is_reply'] = df['reply_timestamp'].notna()
df['is_like'] = df['like_timestamp'].notna()
df['is_retweet_with_comment'] = df['retweet_with_comment_timestamp'].notna()
df['is_retweet'] = df['retweet_timestamp'].notna()

In [27]:
df[['reply_timestamp','is_reply', 'like_timestamp', 'is_like', 'retweet_with_comment_timestamp', 'is_retweet_with_comment','retweet_timestamp' ,'is_retweet']].head(20)

reply_timestamp  is_reply  like_timestamp  is_like  \
0               NaN     False             NaN    False   
1               NaN     False             NaN    False   
2               NaN     False             NaN    False   
3               NaN     False    1.581305e+09     True   
4               NaN     False             NaN    False   
5               NaN     False             NaN    False   
6               NaN     False             NaN    False   
7               NaN     False    1.581548e+09     True   
8               NaN     False             NaN    False   
9               NaN     False    1.581548e+09     True   
10              NaN     False             NaN    False   
11              NaN     False    1.581089e+09     True   
12              NaN     False    1.581250e+09     True   
13              NaN     False             NaN    False   
14              NaN     False    1.581299e+09     True   
15     1.580951e+09      True             NaN    False   
16              NaN     False             NaN    False   
17              NaN     False    1.581328e+09     True   
18              NaN     False    1.581156e+09     True   
19              NaN     False             NaN    False   

    retweet_with_comment_timestamp  is_retweet_with_comment  \
0                              NaN                    False   
1                              NaN                    False   
2                              NaN                    False   
3                              NaN                    False   
4                              NaN                    False   
5                              NaN                    False   
6                              NaN                    False   
7                              NaN                    False   
8                              NaN                    False   
9                              NaN                    False   
10                             NaN                    False   
11                             NaN                    False   
12                             NaN                    False   
13                             NaN                    False   
14                             NaN                    False   
15                             NaN                    False   
16                             NaN                    False   
17                             NaN                    False   
18                             NaN                    False   
19                             NaN                    False   

    retweet_timestamp  is_retweet  
0                 NaN       False  
1                 NaN       False  
2                 NaN       False  
3                 NaN       False  
4                 NaN       False  
5                 NaN       False  
6                 NaN       False  
7                 NaN       False  
8                 NaN       False  
9                 NaN       False  
10       1.581157e+09        True  
11                NaN       False  
12                NaN       False  
13                NaN       False  
14                NaN       False  
15                NaN       False  
16                NaN       False  
17                NaN       False  
18                NaN       False  
19                NaN       False

In [28]:
display(df.head())

text_tokens  \
0  101\t56898\t137\t174\t63247\t10526\t131\t3197\...   
1  101\t102463\t10230\t10105\t21040\t10169\t12811...   
2  101\t56898\t137\t11255\t22037\t10263\t168\t111...   
3  101\t13073\t28757\t106\t100\t14120\t131\t120\t...   
4  101\t3460\t1923\t6632\t2824\t30368\t2179\t1881...   

                           hashtags                          tweet_id  \
0                               NaN  3C21DCFB8E3FEC1CB3D2BFB413A78220   
1                               NaN  3D87CC3655C276F1771752081423B405   
2  DB32BD91C2F1B37BE700F374A07FBC61  3701848B96AA740528A2B0E247777D7D   
3                               NaN  18176C6AD2871729384062F073CCE94D   
4                               NaN  AF11AF01F842E7F120667B7B0B38676D   

  present_media                     present_links  \
0         Video                               NaN   
1           NaN  BB422AA00380E45F312FD2CAA75F4960   
2           NaN  2423BA02A75DB2189335DDC3FB6B74A1   
3         Video                               NaN   
4           NaN                               NaN   

                    present_domains tweet_type  \
0                               NaN    Retweet   
1  92D397F8E0F1E77B36B8C612C2C51E23   TopLevel   
2  6D323BE93766E79BE423FAC5C28BE39B    Retweet   
3                               NaN   TopLevel   
4                               NaN      Quote   

                           language  tweet_timestamp  \
0  76B8A9C3013AE6414A3E6012413CDC3B       1581467323   
1  D3164C7FBCF2565DDF915B1B3AEFB1DC       1580975391   
2  22C448FF81263D4BAF2A176145EE9EAD       1581257232   
3  D3164C7FBCF2565DDF915B1B3AEFB1DC       1581164292   
4  22C448FF81263D4BAF2A176145EE9EAD       1581233650   

               engaged_with_user_id  engaged_with_user_follower_count  \
0  D1AA2C85FA644D64346EDD88470525F2                               737   
1  4DC65AC7BD963DE1F7617C047C33DE99                          52366425   
2  5C671539CB41B9807E209349B101E9FF                               988   
3  70B900BE17416923D1E236A38798F202                           1228134   
4  E94C0E9E8494F3D603F9D1A5C5242E3D                                73   

   engaged_with_user_following_count  engaged_with_user_is_verified  \
0                                706                          False   
1                               2383                           True   
2                                167                          False   
3                               5413                          False   
4                                299                          False   

   engaged_with_user_account_creation                  engaging_user_id  \
0                          1403069820  000046C8606F1C3F5A7296222C88084B   
1                          1230139136  00006047187D0D18598EF12A650E1DAC   
2                          1530094483  0000648BAA193AE4C625DDF789B57172   
3                          1378699943  000071667F50BAFEA722A8E8284581E5   
4                          1549054499  00007745A6EE969F1A0F44B10DC17671   

   engaging_user_follower_count  engaging_user_following_count  \
0                           131                           2105   
1                            22                             50   
2                           251                            719   
3                            18                             58   
4                           268                            526   

   engaging_user_is_verified  engaging_user_account_creation  \
0                      False                      1573978269   
1                      False                      1340673962   
2                      False                      1456473671   
3                      False                      1378427564   
4                      False                      1252294800   

   engaged_follows_engaging  reply_timestamp  retweet_timestamp  \
0                     False              NaN                NaN   
1                     False              NaN                NaN   
2       

### T4. Extract the Social Network
* Twitter Social Network is directional (follower – following) 
* parse the `engaged_follows_engaging` field: each example gives you an edge
* Create the adjacency matrix representation of the social graph
* If an edge exists between two users, 0 otherwise
* how can you use this information?

In [29]:
# Adjacency matrix

## Format data

**Task: Format data**\
Formatting transformations refer to primarily syntactic modifications
made to the data that do not change its meaning, but might be required
by the modeling tool.


**Output: Reformatted data**\
Some tools have requirements on the order of the attributes, such as
the first field being a unique identifier for each record or the last field
being the outcome field the model is to predict.
It might be important to change the order of the records in the dataset.
Perhaps the modeling tool requires that the records be sorted according
to the value of the outcome attribute. A common situation is that the
records of the dataset are initially ordered in some way but the modeling
algorithm needs them to be in a fairly random order. For example, when
using neural networks it is generally best for the records to be presented
in a random order although some tools handle this automatically without explicit user intervention

In [30]:
# Parse attributes containing tab-separated lists into lists.
df['text_tokens'] = df['text_tokens'].str.split('\t')

def to_hex_list(x):
    output = str(x).split('\t')
#     output = [int(val, 16) for val in str(x).split('\t')] 
    return output

cols_to_process = ['hashtags', 'present_media', 'present_links', 'present_domains']

for col in cols_to_process:  
    df[col] = df[col].apply(lambda x: to_hex_list(x) if isinstance(x, str)  else x)


# Transform raw timestamps into human-readable timestamps.
cols_to_process = ['tweet_timestamp', 'engaging_user_account_creation', 'reply_timestamp', 'retweet_timestamp', 'retweet_with_comment_timestamp', 'like_timestamp']

for col in cols_to_process:  
    df[col] = df[col].apply(lambda x: pd.Timestamp(x, unit='s'))

# Modeling

## T1. Split into train, dev, test
* Sub-sample to create test, non-test datasets
* Optionally split non-test into train and dev
* e.g., to implement k-fold validation

In [31]:
# Train/Test split

from sklearn.model_selection import train_test_split
X_nontest, X_test = train_test_split(df, test_size = 0.1, random_state = 42)
X_train, X_dev = train_test_split(X_nontest, test_size = 0.1, random_state = 42)
print(X_train.shape, X_dev.shape, X_test.shape)

(65143, 37) (7239, 37) (8043, 37)


## Naive Bayes 

## Multivariate Regression

## Neural network approach

### T5. Implement a Baseline
* Implement the neural network approach described in the challenge
paper: https://arxiv.org/abs/2004.13715

# Evaluation

## Eevaluate as a binary classification task in two ways
* Area Under Precision-Recall Curve
  * generate precision-recall pairs for various probability thresholds
    * assumes anything above threshold is predicted as relevant
* Cross-Entropy Loss = Log-Loss (for binary classification)
  * measure how good the predicted probabilities are

### T2.1 Parse test to create the ground truth output file
engaging user id; tweet id; label


In [32]:
# gt.csv

### T2.2: Implement the `read_predictions` function 
from https://recsys-twitter.com/code/snippets

In [33]:
from sklearn.metrics import precision_recall_curve, auc, log_loss

def compute_prauc(pred, gt):
  prec, recall, thresh = precision_recall_curve(gt, pred)
  prauc = auc(recall, prec)
  return prauc

def calculate_ctr(gt):
  positive = len([x for x in gt if x == 1])
  ctr = positive/float(len(gt))
  return ctr

def compute_rce(pred, gt):
    cross_entropy = log_loss(gt, pred)
    data_ctr = calculate_ctr(gt)
    strawman_cross_entropy = log_loss(gt, [data_ctr for _ in range(len(gt))])
    return (1.0 - cross_entropy/strawman_cross_entropy)*100.0

  
ground_truth = read_predictions("gt.csv") # will return data in the form (tweet_id, user_id, labed (1 or 0))
predictions = read_predictions("predictions.csv") # will return data in the form (tweet_id, user_id, prediction)

NameError: name 'read_predictions' is not defined